File: 0155_Min_Stack/155.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.L = []

    def push(self, val: int) -> None:
        self.L.append(val)

    def pop(self) -> None:
        self.L.pop()

    def top(self) -> int:
        return self.L[-1]

    def getMin(self) -> int:
        return min(self.L)


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()


# Input
# ["MinStack","push","push","push","getMin","pop","top","getMin"]
# [[],[-2],[0],[-3],[],[],[],[]]

# Output
# [null,null,null,null,-3,null,0,-2]

result = []
obj = MinStack()
result.append(None)
result.append(obj.push(-2))
result.append(obj.push(0))
result.append(obj.push(-3))
result.append(obj.getMin())
result.append(obj.pop())
result.append(obj.top())
result.append(obj.getMin())

print(result)

File: 0155_Min_Stack/155.cpp
#include <bits/stdc++.h>
using namespace std;

class MinStack
{
private:
    vector<int> stk;

public:
    /** initialize your data structure here. */
    MinStack()
    {
        stk.clear();
    }

    void push(int val)
    {
        stk.push_back(val);
    }

    void pop()
    {
        stk.pop_back();
    }

    int top()
    {
        return stk[stk.size() - 1];
    }

    int getMin()
    {
        return *min_element(stk.begin(), stk.end());
        // int m = stk[0];
        // for (int i = 1; i < stk.size(); i++)
        // {
        //     if (stk[i] < m)
        //         m = stk[i];
        // }
        // return m;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */